src/model/timestamp.rs

```diff
@@ -0,0 +1,184 @@
+///
+/// The struct implements the `std::fmt::Display` trait to format the underlying type as
+/// an RFC 3339 date and string such as `2016-04-30T11:18:25.796Z`.
+///
+/// ```
+/// # use serenity::model::id::GuildId;
+/// # use serenity::model::Timestamp;
+/// #
+/// let timestamp: Timestamp = GuildId(175928847299117063).created_at();
+/// assert_eq!(timestamp.unix_timestamp(), 1462015105);
+/// assert_eq!(timestamp.to_string(), "2016-04-30T11:18:25.796Z");
+/// ```
+#[cfg(not(feature = "time"))]
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct Timestamp(DateTime<Utc>);
+
+/// Representation of a Unix timestamp.
+///
+/// The struct implements the `std::fmt::Display` trait to format the underlying type as
+/// an RFC 3339 date and string such as `2016-04-30T11:18:25.796Z`.
+///
+/// ```
+/// # use serenity::model::id::GuildId;
+/// # use serenity::model::Timestamp;
+/// #
+/// let timestamp: Timestamp = GuildId(175928847299117063).created_at();
+/// assert_eq!(timestamp.unix_timestamp(), 1462015105);
+/// assert_eq!(timestamp.to_string(), "2016-04-30T11:18:25.796Z");
+/// ```
+#[cfg(feature = "time")]
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct Timestamp(#[serde(with = "rfc3339")] OffsetDateTime);
+
+#[cfg(feature = "time")]
+impl Timestamp {
+    pub(crate) fn from_discord_id(id: u64) -> Timestamp {
+        let ns = Duration::milliseconds(((id >> 22) + DISCORD_EPOCH) as i64).whole_nanoseconds();
+        // This can't fail because of the bit shifting
+        // `(u64::MAX >> 22) + DISCORD_EPOCH` = 5818116911103 = "Wed May 15 2154 07:35:11 GMT+0000"
+        Self(OffsetDateTime::from_unix_timestamp_nanos(ns).expect("can't fail"))
+    }
+
+    /// Create a new `Timestamp` with the current date and time in UTC.
+    pub fn now() -> Self {
+        Self(OffsetDateTime::now_utc())
+    }
+
+    /// Returns the number of non-leap seconds since January 1, 1970 0:00:00 UTC
+    pub fn unix_timestamp(&self) -> i64 {
+        self.0.unix_timestamp()
+    }
+
+    fn parse(input: &str) -> Result<Timestamp, ParseError> {
+        OffsetDateTime::parse(input, &Rfc3339).map(Self).map_err(|_| ParseError)
+    }
+}
+
+#[cfg(not(feature = "time"))]
+impl Timestamp {
+    pub(crate) fn from_discord_id(id: u64) -> Timestamp {
+        Self(Utc.timestamp_millis(((id >> 22) + DISCORD_EPOCH) as i64))
+    }
+
+    /// Create a new `Timestamp` with the current date and time in UTC.
+    pub fn now() -> Self {
+        Self(Utc::now())
+    }
+
+    /// Returns the number of non-leap seconds since January 1, 1970 0:00:00 UTC
+    pub fn unix_timestamp(&self) -> i64 {
+        self.0.timestamp()
+    }
+
+    fn parse(input: &str) -> Result<Timestamp, ParseError> {
+        DateTime::parse_from_rfc3339(input)
+            .map(|d| Self(d.with_timezone(&Utc)))
+            .map_err(|_| ParseError)
+    }
+}
+
+#[derive(Debug)]
+pub struct ParseError;
```

Comment:
This should have documentation stating what it represents and when it is created. For example:
```rust
/// Signifies the failure to parse the `Timestamp` from an rfc3339 string.
```

Reply:
The `timestamp` module is currently not public but it's good add to it if the module gets exported in the future.